### PR TITLE
controller: validation: check for cpu cores split

### PR DESCRIPTION
**Pull request by fromanirh**
We always assume, and document, that the reserved and
isolated cpus set must not overlap.
Add some validation code to enforce this constraint.

Signed-off-by: Francesco Romani <fromani@redhat.com>

### CODE REVIEW

**Comment by fromanirh** [comment]
/hold
still WIP
missing e2e tests and better commit message

**Comment by cynepco3hahue** [comment]
@fromanirh IIUC it only will prevent the overlaps?(that totally fine for me, because our d/s CI does not include all CPUs under isolated and reserved)

**Comment by fromanirh** [comment]
> @fromanirh IIUC it only will prevent the overlaps?(that totally fine for me, because our d/s CI does not include all CPUs under isolated and reserved)

Yep, only overlaps. This is because other checks are hard/impossible unless we know exactly the worker node HW configuration (how many cores?) and, to some extent, without secondguessing the admin perspective.

TL;DR: overlapping cores is the only wrong setting we can detect reliably.

**Comment by cynepco3hahue** [comment]
@fromanirh Do you need to add something else under the PR?

**Comment by fromanirh** [comment]
> @fromanirh Do you need to add something else under the PR?

yep. e2e test(s).

**Comment by fromanirh** [comment]
/hold cancel
/cc @fedepaol 
Added e2e test and @fedepaol as reviewer because I want to be nice with the discovery mode feature

File: functests/3_performance_status/status.go [hunk 1]
```diff
@@ -133,4 +135,52 @@ var _ = Describe("Status testing of performance profile", func() {
...
+						Enabled: pointer.BoolPtr(true),
+					},
+					NUMA: &performancev1.NUMA{
+						TopologyPolicy: pointer.StringPtr("restricted"),
+					},
+				},
+			}
+			Expect(testclient.Client.Create(context.TODO(), overlappingProfile)).ToNot(HaveOccurred())
+			defer func() {
+				err := testclient.Client.Delete(context.TODO(), overlappingProfile)
```
Comment (fedepaol) [inline functests/3_performance_status/status.go:175]: You should check that the profile is really deleted & that MCP is not updating, otherwise there's a risk that the profile will leak. Check the profile2 test for inspiration

File: functests/3_performance_status/status.go [hunk 2]
```diff
@@ -133,4 +135,52 @@ var _ = Describe("Status testing of performance profile", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+			}()
+
+			nodeLabels := map[string]string{
+				newLabel: "",
+			}
+			condMessage := profiles.GetConditionMessage(nodeLabels, v1.ConditionDegraded)
```
Comment (fedepaol) [inline functests/3_performance_status/status.go:182]: Should this be wrapped by an eventually?

**Review by fedepaol — COMMENTED** [review]
A couple of comments

File: functests/3_performance_status/status.go (hunk 1 above)
Comment (fromanirh) [reply functests/3_performance_status/status.go:175]: Ack. Do we have a way to recover from a leaked profile?

File: functests/3_performance_status/status.go (hunk 2 above)
Comment (fromanirh) [reply functests/3_performance_status/status.go:182]: Probably better, yes.

File: functests/3_performance_status/status.go (hunk 1 above)
Comment (fedepaol) [reply functests/3_performance_status/status.go:175]: Not sure I understand the question, just make sure that when the test exits the profile is really deleted (as we do with the other profile2 test)

Comment (fromanirh) [reply functests/3_performance_status/status.go:175]: Sorry, I was too terse. I got it about making sure the profile should not leak. Followup question: should a profile actually leak (because someone like @fromanirh added a broken test!), do we have a way to recover or this means we polluted the tested cluster and the only remedy is let the cluster admin know so it can be fixed manually?

Comment (fedepaol) [reply functests/3_performance_status/status.go:175]: It's the tests responsability not to leave leftovers. In many cases this is achieved by cleaning the namespaces.
In the case of perf profiles though, the only way to clean them is to explicilty delete them. Whatever mechanism you come with, it will require adding the profile to be deleted somewhere (because you also don't want to delete them all), so it's still the dev's responsibility to make sure state don't leak (as with all the other cases).

Comment (fromanirh) [reply functests/3_performance_status/status.go:175]: Got it and makes sense. Thanks!

**Comment by fromanirh** [comment]
ok, looks like this failure is real: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift-kni_performance-addon-operators/378/pull-ci-openshift-kni-performance-addon-operators-master-e2e-gcp-operator-upgrade/1313366629148528640

I'm not completely sure why the e2e-gcp lane passed cleanly and the upgrade lane is not, so this deserve some extra investigation

File: functests/3_performance_status/status.go [hunk 3]
```diff
@@ -133,4 +136,66 @@ var _ = Describe("Status testing of performance profile", func() {
...
+
+			}()
+
+			nodeLabels := map[string]string{
+				newLabel: "",
+			}
+
+			var cond *v1.Condition
+			Eventually(func() bool {
+				cond = profiles.GetCondition(nodeLabels, v1.ConditionDegraded)
```
Comment (fedepaol) [inline functests/3_performance_status/status.go:192]: I had a hard time understanding what this does. Would it be possible to have a local "isProfileDegraded" function (or something along that line) wrapping the code?

File: pkg/controller/performanceprofile/components/utils.go [hunk 4]
```diff
@@ -108,3 +108,18 @@ func CPUListToMaskList(cpulist string) (hexMask string, err error) {
 	trimmedCPUMaskList := b.String()
 	return trimmedCPUMaskList, nil
 }
+
+// CPUListIntersect returns cpu ids found in both the provided cpulists, if any
+func CPUListIntersect(cpulistA, cpulistB string) ([]int, error) {
```
Comment (fedepaol) [inline pkg/controller/performanceprofile/components/utils.go:113]: nit: cpuList

**Review by fedepaol — COMMENTED** [review]
Just a couple of nits, looks good!
